### PR TITLE
Refactor Client Data Source Interfaces

### PR DESCRIPTION
**Pull request by matteblair**
The new interfaces should substantially reduce the CPU workload for adding large sets of features and should be more clear to understand.

Using these interfaces for data sets of any substantial size is currently very inefficient for several reasons.
1. Assigning content to the data sources can only be done by adding features one at a time and the entire contents of the source is re-tiled with each addition. This makes adding N features a roughly O(N^2) operation.
2. Feature data goes through redundantly copies in order to conform to expected types at interface boundaries. This involves allocating and looping over the data each time, with no material benefit.

This PR addresses 1 by changing the way content is assigned to the data source. The new interface requires you to assign the entire list of features _or_ the entire geoJSON contents at once, and then tiles the data. The application can retain their list of features, manipulate it however they want, and assign it again to update the data source.

2 is addressed by indirectly exposing the final geometry data structure needed by the data source via a pair of "builder" structs - this eliminates the need to allocate and loop over intermediate data structures.

This will be a breaking change to the Map Data interfaces, but definitely a necessary one.

Follow-up work:
- Update geojson-vt-cpp version - new versions enable more efficient construction of features.
- Refactor `com.mapzen.tangram.geometry` classes - I don't know what I was thinking when I implemented these, geometry should be separate from feature properties and should not transform its own data into arrays until needed.

Addresses https://github.com/tangrams/tangram-es/issues/1694, https://github.com/tangrams/tangram-es/issues/1135
Related to https://github.com/tangrams/tangram-es/issues/1565

### CODE REVIEW

**Comment by matteblair** [comment]
Just to verify that this _does_ in fact solve the performance bottlenecks I was going after, I wrote a little code to measure how long it takes to add 1000 points to a `TGMapData` before and after this change. 

Old code patch:
```diff
diff --git a/platforms/ios/demo/src/MapViewController.m b/platforms/ios/demo/src/MapViewController.m
index da3275b71..1021c9d8e 100644
--- a/platforms/ios/demo/src/MapViewController.m
+++ b/platforms/ios/demo/src/MapViewController.m
@@ -56,8 +56,7 @@ - (void)mapView:(TGMapView *)mapView didLoadScene:(int)sceneID withError:(nullab
         return;
     }
 
-    // Add a client data source, named 'mz_route_line_transit'
-    self.mapData = [mapView addDataLayer:@"mz_route_line_transit" generateCentroid:NO];
+    self.mapData = [mapView addDataLayer:@"mz_default_point" generateCentroid:NO];
 }
 
 - (void)mapView:(TGMapView *)mapView didSelectMarker:(TGMarkerPickResult *)markerPickResult atScreenPosition:(CGPoint)position;
@@ -170,6 +169,20 @@ - (void)mapView:(TGMapView *)view recognizer:(UIGestureRecognizer *)recognizer d
 - (void)mapView:(TGMapView *)mapView recognizer:(UIGestureRecognizer *)recognizer didRecognizeLongPressGesture:(CGPoint)location
 {
     NSLog(@"Did long press at %f %f", location.x, location.y);
+    CFTimeInterval startTime = CACurrentMediaTime();
+    const int NUM_POINTS = 1000;
+    TGFeatureProperties *properties = @{ @"type" : @"point", @"color" : @"#D2655F" };
+    CLLocationCoordinate2D center = [mapView coordinateFromViewPosition:location];
+    double theta = 0, radius = 0;
+    for (int i = 0; i < NUM_POINTS; i++) {
+        theta += 30.0 / NUM_POINTS;
+        radius += .1 / NUM_POINTS;
+        CLLocationCoordinate2D point = CLLocationCoordinate2DMake(center.latitude + radius * sin(theta), center.longitude + radius * cos(theta));
+        [_mapData addPoint:point withProperties:properties];
+    }
+    CFTimeInterval endTime = CACurrentMediaTime();
+    NSLog(@"Updated MapData in %f seconds", endTime - startTime);
+    [mapView requestRender];
 }
```

New code patch:
```diff
diff --git a/platforms/ios/demo/src/MapViewController.m b/platforms/ios/demo/src/MapViewController.m
index 0f8487a0d..2602f2189 100644
--- a/platforms/ios/demo/src/MapViewController.m
+++ b/platforms/ios/demo/src/MapViewController.m
@@ -61,8 +61,7 @@ - (void)mapView:(TGMapView *)mapView didLoadScene:(int)sceneID withError:(nullab
         return;
     }
 
-    // Add a client data source, named 'mz_route_line_transit'
-    self.mapData = [mapView addDataLayer:@"mz_route_line_transit" generateCentroid:NO];
+    self.mapData = [mapView addDataLayer:@"mz_default_point" generateCentroid:NO];
 }
 
 - (void)mapView:(TGMapView *)mapView didSelectMarker:(TGMarkerPickResult *)markerPickResult atScreenPosition:(CGPoint)position;
@@ -167,6 +166,23 @@ - (void)mapView:(TGMapView *)view recognizer:(UIGestureRecognizer *)recognizer d
 - (void)mapView:(TGMapView *)mapView recognizer:(UIGestureRecognizer *)recognizer didRecognizeLongPressGesture:(CGPoint)location
 {
     NSLog(@"Did long press at %f %f", location.x, location.y);
+
+    CFTimeInterval startTime = CACurrentMediaTime();
+    const int NUM_POINTS = 1000;
+    TGFeatureProperties *properties = @{ @"type" : @"point", @"color" : @"#D2655F" };
+    NSMutableArray<TGMapFeature *> *features = [NSMutableArray arrayWithCapacity:NUM_POINTS];
+    CLLocationCoordinate2D center = [mapView coordinateFromViewPosition:location];
+    double theta = 0, radius = 0;
+    for (int i = 0; i < NUM_POINTS; i++) {
+        theta += 30.0 / NUM_POINTS;
+        radius += .1 / NUM_POINTS;
+        CLLocationCoordinate2D point = CLLocationCoordinate2DMake(center.latitude + radius * sin(theta), center.longitude + radius * cos(theta));
+        [features addObject:[TGMapFeature mapFeatureWithPoint:point properties:properties]];
+    }
+    [_mapData setFeatures:features];
+    CFTimeInterval endTime = CACurrentMediaTime();
+    NSLog(@"Updated MapData in %f seconds", endTime - startTime);
+    [mapView requestRender];
 }
 ```

Output on map:
screen shot 2019-02-26 at 7 16 19 pm

Time taken in old `TGMapData`: __4.511 seconds__
Time taken in new `TGMapData`: __0.022 seconds__

**Comment by tallytalwar** [comment]
Sorry got late with this. Will wrap up my review for this today.

File: platforms/ios/framework/src/TGMapFeature+Internal.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+//  Created by Matt Blair on 2/25/19.
+//
+
+#import "TGMapFeature.h"
+
+#pragma mark Common Map Feature Behavior
+
+@interface TGMapFeature ()
+
+- (instancetype)initWithProperties:(TGFeatureProperties *)properties;
```
Comment (tallytalwar) [inline platforms/ios/framework/src/TGMapFeature+Internal.h]: Whats the use case for a map feature without any associated geo location but just properties?

Comment (tallytalwar) [reply platforms/ios/framework/src/TGMapFeature+Internal.h]: Ok, I see its used internally to instantiate parent `TGMapFeature`: https://github.com/tangrams/tangram-es/blob/53b302197e0325f84b2f0525193c7ffde23942bb/platforms/ios/framework/src/TGMapFeature.m#L61

File: platforms/ios/framework/src/TGMapData.mm [hunk 2]
```diff
@@ -54,86 +50,60 @@ - (instancetype)initWithMapView:(__weak TGMapView *)mapView name:(NSString *)nam
...
-- (void)addGeoJson:(NSString *)data
+- (void)setGeoJson:(NSString *)data
 {
     if (!self.map) {
         return;
     }
 
     std::string sourceData = std::string([data UTF8String]);
+    dataSource->clearData();
     dataSource->addData(sourceData);
```
[the remark this answers is not in the document]
Comment (matteblair) [reply platforms/ios/framework/src/TGMapData.mm]: Ah yep, good call.

[the remark this answers is not in the document]
Comment (tallytalwar) [reply platforms/ios/framework/src/TGMapData.mm]: Yup tested with the following geojson:
```json
{
  "type": "Feature",
  "properties": {
    "name": "abc"
  },
  "geometry": {
    "type": "LineString",
    "coordinates": [
      [
        -122.48369693756104,
        37.83381888486939
      ],
      [
        -122.48348236083984,
        37.83317489144141
      ],
      [
        -122.48339653015138,
        37.83270036637107
      ],
      [
        -122.48356819152832,
        37.832056363179625
      ],
      [
        -122.48404026031496,
        37.83114119107971
      ],
      [
        -122.48404026031496,
        37.83049717427869
      ],
      [
        -122.48348236083984,
        37.829920943955045
      ],
      [
        -122.48356819152832,
        37.82954808664175
      ],
      [
        -122.48507022857666,
        37.82944639795659
      ],
      [
        -122.48610019683838,
        37.82880236636284
      ],
      [
        -122.48695850372314,
        37.82931081282506
      ],
      [
        -122.48700141906738,
        37.83080223556934
      ],
      [
        -122.48751640319824,
        37.83168351665737
      ],
      [
        -122.48803138732912,
        37.832158048267786
      ],
      [
        -122.48888969421387,
        37.83297152392784
      ],
      [
        -122.48987674713133,
        37.83263257682617
      ],
      [
        -122.49043464660643,
        37.832937629287755
      ],
      [
        -122.49125003814696,
        37.832429207817725
      ],
      [
        -122.49163627624512,
        37.832564787218985
      ],
      [
        -122.49223709106445,
        37.83337825839438
      ],
      [
        -122.49378204345702,
        37.83368330777276
      ]
    ]
  }
}
```

Not having `generateTiles` does not draw the route lines at all. But with the fix from 889e9db fixes it.

**Pull request merged**